connector/connect/server/src/main/scala/org/apache/spark/sql/connect/service/SparkConnectSessionManager.scala

```diff
@@ -95,47 +92,134 @@ class SparkConnectSessionManager extends Logging {
+      executor.shutdown()
+      executor.awaitTermination(1, TimeUnit.MINUTES)
+    }
+    scheduledExecutor = None
+    // note: this does not cleanly shut down the sessions, but the server is shutting down.
+    sessionStore.clear()
+    closedSessionsCache.invalidateAll()
+  }
+
+  def listActiveSessions: Seq[SessionHolderInfo] = sessionsLock.synchronized {
+    sessionStore.values.map(_.getSessionHolderInfo).toSeq
+  }
+
+  def listClosedSessions: Seq[SessionHolderInfo] = sessionsLock.synchronized {
+    closedSessionsCache.asMap.asScala.values.toSeq
+  }
+
+  /**
+   * Schedules periodic maintenance checks if it is not already scheduled.
+   *
+   * The checks are looking to remove sessions that expired.
+   */
+  private def schedulePeriodicChecks(): Unit = sessionsLock.synchronized {
+    scheduledExecutor match {
+      case Some(_) => // Already running.
+      case None =>
+        val interval = SparkEnv.get.conf.get(CONNECT_SESSION_MANAGER_MAINTENANCE_INTERVAL).toLong
+        // Default for sessions that don't have expiration time set.
+        val defaultInactiveTimeout =
+          SparkEnv.get.conf.get(CONNECT_SESSION_MANAGER_DEFAULT_SESSION_TIMEOUT).toLong
+        logInfo(s"Starting thread for cleanup of expired sessions every $interval ms")
+        scheduledExecutor = Some(Executors.newSingleThreadScheduledExecutor())
+        scheduledExecutor.get.scheduleAtFixedRate(
+          () => {
+            try periodicMaintenance(defaultInactiveTimeout)
+            catch {
+              case NonFatal(ex) => logWarning("Unexpected exception in periodic task", ex)
+            }
+          },
+          interval,
+          interval,
+          TimeUnit.MILLISECONDS)
+    }
+  }
+
+  // Visible for testing.
+  private[connect] def periodicMaintenance(
+      defaultInactiveTimeout: Long,
+      ignoreExpirationTime: Boolean = false): Unit = {
+    logInfo("Started periodic run of SparkConnectSessionManager maintenance.")
+    // Find any sessions that expired and should be removed.
+    val toRemove = new mutable.ArrayBuffer[SessionHolder]()
     sessionsLock.synchronized {
-      sessionStore.invalidateAll()
-      closedSessionsCache.invalidateAll()
+      val nowMs = System.currentTimeMillis()
+
+      sessionStore.values.foreach { sessionHolder =>
+        val info = sessionHolder.getSessionHolderInfo
+        if (info.customExpirationTime.isDefined && !ignoreExpirationTime) {
+          // Custom expirationTime set.
+          if (info.customExpirationTime.get <= nowMs) {
+            toRemove += sessionHolder
+          }
+        } else {
+          // Default inactive timeout from last access.
+          if (info.lastAccessTime + defaultInactiveTimeout <= nowMs) {
+            toRemove += sessionHolder
+          }
+        }
+      }
+    }
+    if (!toRemove.isEmpty) {
+      // .. and remove them.
+      toRemove.foreach { sessionHolder =>
+        val info = sessionHolder.getSessionHolderInfo
```

Comment:
Should we check if this session had its `lastActivityTime` updated?

Reply:
What is the resolution here?

Reply:
> What is the resolution here?

The updated code checks the last activity once again now, giving the session one last chance to get last moment activity.